tests/test_async_actor.rs

```diff
@@ -0,0 +1,67 @@
+use std::pin::Pin;
+
+use futures::{future::ready, StreamExt};
+
+use actix::prelude::*;
+use actix::{AsyncHandler, AsyncMessage, TempRef};
+use tokio::time::delay_for;
+use std::time::{Duration, Instant};
+
+#[derive(Debug, Clone, Copy)]
+struct Num(usize);
+
+impl Message for Num {
+    type Result = usize;
+}
+
+impl AsyncMessage for Num {}
+
+struct MyActor {
+    count: usize,
+}
+
+impl Actor for MyActor {
+    type Context = actix::Context<Self>;
+}
+
+impl AsyncHandler<Num> for MyActor {
+    type Result = Pin<Box<dyn Future<Output = <Num as Message>::Result>>>;
+
+    fn handle(
+        mut this: TempRef<Self>,
+        msg: Num,
+        _ctx: TempRef<Self::Context>,
+    ) -> Self::Result {
+        let res = tokio::task::spawn(async move {
```

Comment:
If we change this block to:
```Rust
        mut ctx: TempRef<Self::Context>,
    ) -> Self::Result {
        ctx.as_mut().spawn(
            actix::fut::wrap_future(async {})
                .then(move |_, act: &mut Self, _| {
                    act.count += msg.0;
                    delay_for(Duration::from_millis(msg.0 as u64)).into_actor(act)
                })
                .map(move |_, act: &mut Self, _| {
                    act.count -= msg.0;
                }),
        );
        let res = tokio::task::spawn(async move {
```

I think it should be easy to see that the spawned future binded at `res` does not have exclusive access ove actor's internal state and this can lead to unpredictable behavior, so I don't think this implementation correctly follows Actor's semantic.

This happens because the Runtime, not the Actor's Context, is handling the futures that have access over the Actor's self reference and context.

Reply:
Ok, I see.
But I mentioned it somehow in comments, that maybe 'context' interface should be minified, in order to avoid such problems. If we expose only 'CTX.address()' method in 'TempRef<Context>' then this problem won't exist.
Personally, I don't see a reason to use 'spawn' method if you can write regular async code. 
If something needs sceduling, then we can have a separate method for this.

Reply:
One idea, how this can be solved: with current approach: instead of calling `Self::handle(actor, msg, tmp_ctx);` in sync code we can do this:
```
let fut_res = async move {
    Self::handle(actor, msg, tmp_ctx).await
};

let (s, r) = oneshot::channel::<M::Result>();

ctx.wait(wrap_future(async move {
    let res = fut_res.await;
    s.send(res)
        .unwrap_or_else(|_| panic!("Failed to send future result"));
}));
```
Then we can be sure, that the message processing is already blocked for this actor and `ctx.as_mut().spawn(...);` won't be executed until we finish processing.

Reply:
Implemented in 29296c7. Does it solve it?